### PR TITLE
Add more APIs to configuration input tracking

### PR DESCRIPTION
**Pull request by h0tk3y**
⚠️ Note for reviewers:

The interceptor declarations here use the tool introduced in https://github.com/gradle/gradle/pull/24152.
To see the code generated by the tool, run `:core:compileJava` and then see:
* `subprojects/core/build/generated/sources/annotationProcessor/java/main/org/gradle/internal/classpath/InterceptorDeclaration_GroovyInterceptorsImpl.java`
* `subprojects/core/build/generated/sources/annotationProcessor/java/main/org/gradle/internal/classpath/InterceptorDeclaration_JvmBytecodeImpl.java`

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

File: subprojects/core/src/main/java/org/gradle/internal/classpath/declarations/GroovyFileInterceptors.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    public static String intercept_getText(
+        @Receiver File self,
+        String charset,
+        @CallerClassName String consumer
+    ) throws IOException {
+        return Instrumented.groovyFileGetText(self, charset, consumer);
+    }
+
+    @InterceptCalls
+    @StaticMethod(ofClass = ResourceGroovyMethods.class)
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/declarations/GroovyFileInterceptors.java:59]: 👍

File: subprojects/core/src/main/java/org/gradle/internal/classpath/declarations/KotlinFileExtensionsInterceptorsDeclaration.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    @InterceptCalls
+    @StaticMethod(ofClass = FilesKt.class)
+    public static String intercept_readText(
+        File thisFile,
+        Charset charset,
+        @KotlinDefaultMask int mask,
+        @CallerClassName String consumer
+    ) throws Throwable {
+        return mask != 0 ?
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/declarations/KotlinFileExtensionsInterceptorsDeclaration.java:44]: The plot thickens. 👍

File: subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/model/CallableKindInfo.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+package org.gradle.internal.instrumentation.model;
+
+import org.gradle.internal.instrumentation.api.annotations.CallableKind;
+
+import java.lang.annotation.Annotation;
+
+public enum CallableKindInfo {
+    STATIC_METHOD, INSTANCE_METHOD, AFTER_CONSTRUCTOR, GROOVY_PROPERTY;
+
+    public static CallableKindInfo fromAnnotation(Annotation annotation) {
```
Comment (bamboo) [inline subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/model/CallableKindInfo.java:26]: 👍

**Review by bamboo — CHANGES_REQUESTED** [review]
I haven't reviewed the generators in detail yet, will do it in a second pass but overall this is looking really good.

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Comment by h0tk3y** [comment]
I have updated this PR to include only the changes in interceptors but not the tool itself.
The PR for the tool is https://github.com/gradle/gradle/pull/24152

File: subprojects/core/src/main/java/org/gradle/internal/classpath/FileUtils.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+     * Checks if the channel created with the given options will have read access of the file content.
+     * See the contract of {@link Files#newByteChannel(Path, OpenOption...)}.
+     */
+    public static boolean optionsAllowReading(OpenOption[] options) {
+        boolean hasReadMode = false;
+        boolean hasAnotherMode = false;
+
+        for (OpenOption option : options) {
+            if (option == StandardOpenOption.READ) {
+                hasReadMode = true;
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/FileUtils.java]: Fixed in [`9866fbc` (#23925)](https://github.com/gradle/gradle/pull/23925/commits/9866fbc27ebc36f51927ee338d2a8bf89a7baa6b)

File: subprojects/core/build.gradle.kts [hunk 5]
```diff
@@ -182,11 +183,18 @@ dependencies {
...
     crossVersionTestDistributionRuntimeOnly(project(":distributions-core"))
+
+    annotationProcessor(project(":internal-instrumentation-processor"))
+    annotationProcessor(platform(project(":distributions-dependencies")))
 }
 
 strictCompile {
     ignoreRawTypes() // raw types used in public API
 }
+tasks.withType(JavaCompile::class).named("compileJava") {
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/build.gradle.kts]: Fixed, thanks! It's in the other PR now.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 6]
```diff
@@ -109,30 +112,40 @@ public static void discardListener() {
...
+                new LongGetLongInterceptor(),
+                new BooleanGetBooleanInterceptor(),
+                new SystemGetenvInterceptor(),
+                new RuntimeExecInterceptor(),
+                new ProcessGroovyMethodsExecuteInterceptor(),
+                new ProcessBuilderStartInterceptor(),
+                new ProcessBuilderStartPipelineInterceptor()
+            ),
+            StreamSupport.stream(getGeneratedCallInterceptors().spliterator(), false)
+        ).toArray(CallInterceptor[]::new)
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java]: Thanks! Fixed it in the other PR.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 7]
```diff
@@ -109,30 +112,40 @@ public static void discardListener() {
...
+            return Cast.uncheckedCast(callInterceptorsGetter.invoke(null));
+        } catch (ClassNotFoundException | NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
 
 
     // Called by generated code
     @SuppressWarnings("unused")
+
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java]: ✅

File: subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/modelreader/impl/AnnotationCallInterceptionRequestReaderImpl.java [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+            this.reason = reason;
+        }
+    }
+
+    private static final Class<? extends Annotation>[] CALLABLE_KIND_ANNOTATION_CLASSES = Cast.uncheckedNonnullCast(Stream.of(
+        CallableKind.InstanceMethod.class,
+        CallableKind.StaticMethod.class,
+        CallableKind.AfterConstructor.class,
+        CallableKind.GroovyProperty.class
+    ).toArray(Class[]::new));
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/modelreader/impl/AnnotationCallInterceptionRequestReaderImpl.java]: Ah, indeed! Fixed it in the other PR, thanks!

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 9]
```diff
@@ -109,30 +112,40 @@ public static void discardListener() {
...
+                new ProcessBuilderStartInterceptor(),
+                new ProcessBuilderStartPipelineInterceptor()
+            ),
+            StreamSupport.stream(getGeneratedCallInterceptors().spliterator(), false)
+        ).toArray(CallInterceptor[]::new)
+    );
+
+    private static Iterable<CallInterceptor> getGeneratedCallInterceptors() {
+        try {
+            Class<?> generatedClass = Class.forName(InterceptorDeclaration.GROOVY_INTERCEPTORS_GENERATED_CLASS_NAME);
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java]: Yes, it should be simple. I'll do that.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/declarations/FileInterceptorsDeclaration.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+@SuppressWarnings("NewMethodNamingConvention")
+@SpecificJvmCallInterceptors(generatedClassName = InterceptorDeclaration.JVM_BYTECODE_GENERATED_CLASS_NAME)
+@SpecificGroovyCallInterceptors(generatedClassName = InterceptorDeclaration.GROOVY_INTERCEPTORS_GENERATED_CLASS_NAME)
+public class FileInterceptorsDeclaration {
+    @InterceptCalls
+    @InstanceMethod
+    public static File[] intercept_listFiles(
+        @Receiver File thisFile,
+        @CallerClassName String consumer
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/declarations/FileInterceptorsDeclaration.java:39]: Yes, it should be absolutely doable with some support in our bytecode visitors.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java]: Done in the other PR.

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Review by mlopatkin — APPROVED** [review]
Awesome!

**Comment by h0tk3y** [comment]
See the other PR that includes these changes now: https://github.com/gradle/gradle/pull/24165

**Pull request closed without merging**